utxo/src/cache.rs

```diff
@@ -140,16 +140,23 @@ impl<P: UtxosView> UtxosCache<P> {
         tx: &Transaction,
         height: BlockHeight,
     ) -> Result<UtxosTxUndoWithSources, Error> {
-        let (sources, utxos) = tx
+        let sources = tx
             .inputs()
             .iter()
-            .map(|tx_in| {
-                let utxo = self.spend_utxo(tx_in.outpoint())?;
-                Ok((tx_in.outpoint().tx_id(), utxo))
+            .filter_map(|input| match input {
+                TxInput::Utxo(outpoint) => Some(outpoint.tx_id()),
+                TxInput::Account(_) => None,
             })
-            .collect::<Result<Vec<_>, Error>>()?
-            .into_iter()
-            .unzip();
+            .collect();
+
+        let utxos = tx
+            .inputs()
+            .iter()
+            .map(|input| match input {
+                TxInput::Utxo(outpoint) => self.spend_utxo(outpoint).map(Some),
+                TxInput::Account(_) => Ok(None),
+            })
+            .collect::<Result<Vec<_>, Error>>()?;
```

Comment:
Is this change only needed to avoid the intermediate `collect`? Asking this because I've also solved this in my mini-pr https://github.com/mintlayer/mintlayer-core/pull/944 via `itertools::process_results`:
```
let (sources, utxos) = itertools::process_results(
    tx.inputs().iter().map(|tx_in| {
        let utxo = self.spend_utxo(tx_in.outpoint())?;
        Ok((tx_in.outpoint().tx_id(), utxo))
    }),
    |iter| iter.unzip(),
)?;
```
Now I'm wondering which version I should keep.
(I suppose `process_results` should be better, because it iterates over the collection only once and it shouldn't impose any noticeable overhead by itself).

Reply:
The logic has changed. Now `input` has optional outpoint. So `sources` and `utxos` collections can have different length. I removed zipping and split the code into two loops to account for that.